swap/protocol_test.go

```diff
@@ -0,0 +1,297 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"os"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/p2p/protocols"
+	p2ptest "github.com/ethersphere/swarm/p2p/testing"
+)
+
+/*
+TestHandshake creates two mock nodes and initiates an exchange;
+it expects a handshake to take place between the two nodes
+(the handshake would fail because we don't actually use real nodes here)
+*/
+func TestHandshake(t *testing.T) {
+	var err error
+
+	// setup test swap object
+	swap, dir := newTestSwap(t)
+	defer os.RemoveAll(dir)
+	defer swap.Close()
+
+	ctx := context.Background()
+	testDeploy(ctx, swap.backend, swap)
+	// setup the protocolTester, which will allow protocol testing by sending messages
+	protocolTester := p2ptest.NewProtocolTester(swap.owner.privateKey, 2, swap.run)
+
+	// shortcut to creditor node
+	debitor := protocolTester.Nodes[0]
+	creditor := protocolTester.Nodes[1]
+
+	// set balance artifially
+	swap.balances[creditor.ID()] = -42
+
+	// create the expected cheque to be received
+	cheque := newTestCheque()
+
+	// sign the cheque
+	cheque.Signature, err = cheque.Sign(swap.owner.privateKey)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// run the exchange:
+	// trigger a `EmitChequeMsg`
+	// expect HandshakeMsg on each node
+	err = protocolTester.TestExchanges(p2ptest.Exchange{
+		Label: "TestHandshake",
+		Triggers: []p2ptest.Trigger{
```

Comment:
this trigger needs to come after the handshake

Reply:
Yes, I think that this test needs to be adjusted as it results in problems that are noticed with ErrInvalidHandshakeMsg and verifyHandshake.

Reply:
should be solved under #1701.